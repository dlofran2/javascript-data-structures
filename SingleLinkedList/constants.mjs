export const CANNOT_FIND_EMPTY = "Cannot find on empty list";
export const INVALID_INDEX_NEGATIVE = "Invalid index: lists cannot have negative index";
export const INVALID_INDEX_OUTSIDE = "Invalid index: cannot index item outside of list";
export const INVALID_ADD_NEGATIVE = "Invalid index: cannot have a negative index";
export const INVALID_ADD_OUTSIDE = "Invalid index: cannot add item outside in index outside of the list";
export const INVALID_DELETE_NEGATIVE = "Invalid index: cannot delete item at negative index";
export const INVALID_DELETE_OUTSIDE = "Invalid index: cannot delete item outside in index outside of the list";
