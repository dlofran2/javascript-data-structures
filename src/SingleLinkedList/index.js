import { CANNOT_FIND_EMPTY, INVALID_INDEX_NEGATIVE, INVALID_INDEX_OUTSIDE, INVALID_ADD_OUTSIDE, INVALID_ADD_NEGATIVE, INVALID_DELETE_NEGATIVE, INVALID_DELETE_OUTSIDE } from "./constants.js";

class Node {
  constructor(data) {
    this.data = data;
    this._next = null;
  }
}

export default class SingleLinkedList {
  constructor() {
    this.head = null;
  }

  printList() {
    let currentNode = this.head,
      arr = [];

    while (currentNode) {
      console.log(`Data: ${currentNode.data}, Next: ${currentNode._next ? currentNode._next.data : currentNode._next}`);
      arr.push(currentNode.data);
      currentNode = currentNode._next;
    }

    return arr;
  }

  getDataAtPosition(x) {
    let currentNode = this.head,
      count = 0;

    if (this.empty()) {
      throw new Error(CANNOT_FIND_EMPTY);
    } else if (x < 0) {
      throw new Error(INVALID_INDEX_NEGATIVE);
    }

    while (currentNode && count < x) {
      currentNode = currentNode._next;
      count++;
    }

    if (!currentNode) {
      throw new Error(INVALID_INDEX_OUTSIDE);
    }

    return currentNode.data;
  }

  empty() {
    return this.head === null;
  }

  length() {
    let currentNode = this.head,
      count = 0;

    while (currentNode) {
      currentNode = currentNode._next;
      count++;
    }

    return count;
  }

  insertFront(value) {
    if (this.empty()) {
      this.head = new Node(value);
      return;
    }

    const newNode = new Node(value);
    newNode._next = this.head;
    this.head = newNode;
  }

  insertEnd(value) {
    if (this.empty()) {
      this.head = new Node(value);
      return;
    }

    let currentNode = this.head;
    while (currentNode._next !== null) {
      currentNode = currentNode._next;
    }

    currentNode._next = new Node(value);
  }

  insertDataAtPosition(value, index) {
    if (index < 0) {
      throw new Error(INVALID_ADD_NEGATIVE);
    }

    if (index === 0) {
      this.insertFront(value);
      return;
    }

    let currentNode = this.head,
      prevNode = null,
      count = 0;

    while (count < index) {
      if (!currentNode) {
        throw new Error(INVALID_ADD_OUTSIDE);
      }
      prevNode = currentNode;
      currentNode = currentNode._next;
      count++;
    }

    const newNode = new Node(value);
    newNode._next = currentNode;
    prevNode._next = newNode;
  }

  deleteFront() {
    if (this.empty()) return;
    let currentNode = this.head;
    this.head = currentNode._next;
  }

  deleteEnd() {
    if (this.empty()) return;
    let currentNode = this.head,
      prevNode = null;
    while (currentNode && currentNode._next !== null) {
      prevNode = currentNode;
      currentNode = currentNode._next;
    }
    prevNode._next = null;
  }

  deleteDataAtPosition(index) {
    if (index < 0) {
      throw new Error(INVALID_DELETE_NEGATIVE);
    }

    if (index === 0) {
      this.deleteFront(index);
      return;
    }

    let currentNode = this.head,
      prevNode = null,
      count = 0;

    while (count < index) {
      if (!currentNode) {
        throw new Error(INVALID_DELETE_OUTSIDE);
      }
      prevNode = currentNode;
      currentNode = currentNode._next;
      count++;
    }

    prevNode._next = currentNode._next;
  }
}
