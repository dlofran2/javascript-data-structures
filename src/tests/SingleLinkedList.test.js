import SingleLinkedList from "../SingleLinkedList";
import {
  CANNOT_FIND_EMPTY,
  INVALID_INDEX_NEGATIVE,
  INVALID_INDEX_OUTSIDE,
  INVALID_ADD_NEGATIVE,
  INVALID_ADD_OUTSIDE,
  INVALID_DELETE_NEGATIVE,
  INVALID_DELETE_OUTSIDE
} from "../SingleLinkedList/constants";

describe("SingleLinkedList", () => {
  let list;
  beforeEach(() => {
    list = new SingleLinkedList();
  });

  it("should handle initialization", () => {
    expect(list.length()).toBe(0);
    expect(list.empty()).toBeTruthy();
    expect(list.head).toBeNull();
  });

  it("should handle inserting items in the front of the list", () => {
    list.insertFront(200);
    expect(list.getDataAtPosition(0)).toBe(200);
    expect(list.length()).toBe(1);

    list.insertFront(100);
    expect(list.getDataAtPosition(0)).toBe(100);
    expect(list.getDataAtPosition(1)).toBe(200);
    expect(list.length()).toBe(2);
  });

  it("should handle inserting items in the back of the list", () => {
    list.insertEnd(100);
    expect(list.getDataAtPosition(0)).toBe(100);
    expect(list.length()).toBe(1);

    list.insertEnd(200);
    expect(list.getDataAtPosition(0)).toBe(100);
    expect(list.getDataAtPosition(1)).toBe(200);
    expect(list.length()).toBe(2);
  });

  it("should handle inserting items in the middle of the list", () => {
    list.insertEnd(100);
    list.insertEnd(300);
    expect(list.length()).toBe(2);

    list.insertDataAtPosition(200, 1);
    expect(list.getDataAtPosition(0)).toBe(100);
    expect(list.getDataAtPosition(1)).toBe(200);
    expect(list.getDataAtPosition(2)).toBe(300);
    expect(list.length()).toBe(3);
  });

  it("should handle adding items at the front of the list with the insertDataAtPosition method", () => {
    list.insertDataAtPosition(100, 0);
    expect(list.getDataAtPosition(0)).toBe(100);
    expect(list.length()).toBe(1);
  });

  it("should delete items from the front of the list", () => {
    list.insertEnd(100);
    list.insertEnd(200);
    expect(list.length()).toBe(2);

    list.deleteFront();
    expect(list.getDataAtPosition(0)).toBe(200);
    expect(list.length()).toBe(1);

    list.deleteFront();
    expect(list.length()).toBe(0);
  });

  it("should delete items from the end of the list", () => {
    list.insertEnd(100);
    list.insertEnd(200);
    expect(list.length()).toBe(2);

    list.deleteEnd();
    expect(list.getDataAtPosition(0)).toBe(100);
    expect(list.length()).toBe(1);

    list.deleteFront();
    expect(list.length()).toBe(0);
  });

  it("should delete items from the middle of the list", () => {
    list.insertEnd(100);
    list.insertEnd(200);
    list.insertEnd(300);
    expect(list.length()).toBe(3);

    list.deleteDataAtPosition(1);
    expect(list.length()).toBe(2);
    expect(list.getDataAtPosition(0)).toBe(100);
    expect(list.getDataAtPosition(1)).toBe(300);
  });

  it("should handle deleting items at the front of the list with the deleteDataAtPosition method", () => {
    list.insertFront(100);
    expect(list.length()).toBe(1);
    list.deleteDataAtPosition(0);
    expect(list.length()).toBe(0);
  });

  it("should handle attempt to delete on an empty list", () => {
    expect(list.length()).toBe(0);
    list.deleteFront(0);
    list.deleteEnd(0);
  });

  it("should handle error cases of the getDataAtPosition method", () => {
    expect(() => list.getDataAtPosition(0)).toThrowError(new Error(CANNOT_FIND_EMPTY));
    list.insertFront(100);
    expect(() => list.getDataAtPosition(-1)).toThrowError(new Error(INVALID_INDEX_NEGATIVE));
    expect(() => list.getDataAtPosition(2)).toThrowError(new Error(INVALID_INDEX_OUTSIDE));
  });

  it("should handle error cases of the insertDataAtPosition method", () => {
    expect(() => list.insertDataAtPosition(100, -1)).toThrowError(new Error(INVALID_ADD_NEGATIVE));
    expect(() => list.insertDataAtPosition(100, 1)).toThrowError(new Error(INVALID_ADD_OUTSIDE));
  });

  it("should handle error cases of the deleteDataAtPosition method", () => {
    expect(() => list.deleteDataAtPosition(-1)).toThrowError(new Error(INVALID_DELETE_NEGATIVE));
    expect(() => list.deleteDataAtPosition(1)).toThrowError(new Error(INVALID_DELETE_OUTSIDE));
  });

  it("should the printList method", () => {
    list.insertFront(100);
    list.insertFront(200);
    list.insertFront(300);
    list.insertFront(400);
    list.insertFront(500);
    expect(list.printList()).toMatchSnapshot();
  });
});
