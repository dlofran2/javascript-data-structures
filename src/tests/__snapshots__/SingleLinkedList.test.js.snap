// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleLinkedList should the printList method 1`] = `
Array [
  500,
  400,
  300,
  200,
  100,
]
`;
